### PR TITLE
#3434 Actually save sensors etc in the SensorEditPage A1

### LINKED ISSUE #3434: When saving a sensor in SensorEditPage, the sensor should actually be saved
## Is your feature request related to a problem? Please describe.

As per the title!

## Describe the solution you'd like

As per the title!

## Implementation

There is currently a fake callback:

```
const saveSensor = async () => {
    await new Promise(r => setTimeout(r, 3000));
  };
```

Replace this with a real action which would save the new details of the sensor.

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #3434

## Change summary

- Primarily this is hooking up `SensorEditPage` up to use `SensorReducer`, `TemperatureBreachConfigReducer` and `LocationReducer` instead of `SensorDetail`. When you put it in a sentence like this, it doesn't seem 'simpler' ...!
- It adds the functionality that a sensor is actually saved once clicking 'save'. Technically, it will save the temperature breach configs, the location and the sensor, after sending commands to the sensor.
- Can/should be optimized to only save/send commands for what is needed

## Testing

- [ ] Edit a sensor: The location is edited (the sensor code field is really a location code field), and the configs which are related to the location which is related to the sensor is saved.

### Related areas to think about

- Really not sure what to do about this 'hidden' location situation

### CODE REVIEW

File: src/reducers/Entities/LocationReducer.js [hunk 1]
```diff
@@ -1,11 +1,18 @@
 import { LOCATION_ACTIONS } from '../../actions/Entities';
 import { UIDatabase } from '../../database';
+import { ROUTES } from '../../navigation/index';
+
+const plainLocation = (location = {}) => ({
+  id: location.id,
+  description: location.description,
+  code: location.code,
+});
```
Comment (joshxg) [inline src/reducers/Entities/LocationReducer.js:9]: This is the weirdest? part of the PR I think.

- Realm objects can't be spread, so if there was a redux object, trying to edit immutably is a pain.
- Using realm `toJSON` results in an object which has clones of all of the related records also, so if you are to use `update` on the resulting object which has been edited, the potentially stale and related `sensor` would also be saved.
- ONLY using realm objects results in live updates and requires more complex rolling back (as we update after a 'save' button is clicked - this is the same as in the new sensor flow).

So, I decided to take this route, and transform the objects into plain, simple JS objects in the redux store.

Another alternative would be to override `toJSON` in the `Sensor` realm class.

side note: Possibility here to create an abstraction to hide realm implementation such that we use plain JS objects by overriding `toJSON` on all of the realm classes and pass it through the abstraction to be persisted as whatever. Good phase1 of 2 for implementing sqlite.

**Review by mark-prins — CHANGES_REQUESTED** [review]
I saved an existing sensor and the app crashed with `null is not a valid Bluetooth address`
looking into it now.

Also had this warning `Warning: Failed prop type: The prop 'lastSyncDate' is marked as required in 'SensorEditPageComponent', but its value is 'undefined'.`

**Comment by mark-prins** [comment]
@joshxg The method `updateSensor` is expecting a `sensor` parameter, but is getting passed `macAddress, logInterval`

**Comment by joshxg** [comment]
@mark-prins Argh - sorry. I messed this up in the merge.

I changed the params in this commit: [`5ea154e` (#3444)](https://github.com/openmsupply/mobile/pull/3444/commits/5ea154e4161b9322c75be275cfb899680eb544db)

I thought since `updateLogInterval` only required the `macAddress` and `logInterval` and `disable` only required the `macAddress` it might be more explicit to pass those exact params rather than the more obscure 'sensor' object.. but maybe that's silly?

File: src/pages/SensorEditPage.js [hunk 2]
```diff
@@ -239,20 +274,27 @@ const dispatchToProps = (dispatch, ownProps) => {
...
-    dispatch(SensorActions.update(id, 'logInterval', logInterval));
-  const updateDuration = (type, value) =>
-    dispatch(SensorDetailActions.updateConfig(type, 'duration', value * MILLISECONDS.ONE_MINUTE));
-  const updateTemperature = (type, value) =>
-    dispatch(SensorDetailActions.updateConfig(type, 'temperature', value));
-  const saveSensor = async () => {
-    await new Promise(r => setTimeout(r, 3000));
+    dispatch(SensorActions.update(sensorID, 'logInterval', logInterval));
+  const updateDuration = (id, value) =>
+    dispatch(TemperatureBreachConfigActions.update(id, 'duration', value * 60));
```
[the remark this answers is not in the document]
Comment (katherine-sussol) [reply src/pages/SensorEditPage.js:287]: I noticed something a little odd with the duration decrementor as well (I don't think it was necessarily introduced here - but it's just visible in realm now that it's actually saving), in that I'm able to decrement below 0 duration (but shows '1 minute' in the UI). Have been digging around to see if I could find the culprit but maybe better for a separate issue?

File: src/reducers/Entities/LocationReducer.js (hunk 1 above)
Comment (joshxg) [reply src/reducers/Entities/LocationReducer.js:11]: Just something like this in a realm class:

```
toJSON(){
	return { this.id ... }
}
```


Bring on sqlite and working with plain js objects 🤷 

renamed and a brief comment: 064e355

File: src/pages/NewSensor/NewSensorStepThree.js [hunk 3]
```diff
@@ -98,7 +97,7 @@ export const NewSensorStepThreeComponent = ({
           text={vaccineStrings.connect}
           style={{ backgroundColor: SUSSOL_ORANGE }}
           textStyle={{ color: WHITE, textTransform: 'capitalize' }}
-          onPress={() => withLoadingIndicator(connectToSensor(sensor))}
+          onPress={() => withLoadingIndicator(connectToSensor(macAddress, logInterval))}
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/NewSensor/NewSensorStepThree.js]: Sorry! c9e967a

Big merge :(

File: src/pages/SensorEditPage.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/SensorEditPage.js]: you're right - the merge again I tells ya 
b3d6ddf

File: src/reducers/Entities/SensorReducer.js [hunk 4]
```diff
@@ -2,11 +2,21 @@ import { SENSOR_ACTIONS } from '../../actions/Entities/SensorActions';
 import { UIDatabase } from '../../database';
 import { ROUTES } from '../../navigation/index';
 
+const plainSensor = sensor => ({
+  id: sensor.id,
+  macAddress: sensor.macAddress,
+  name: sensor.name,
+  batteryLevel: sensor.batteryLevel,
+  locationID: sensor.location.id,
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/reducers/Entities/SensorReducer.js]: 9a243f8

File: src/pages/NewSensor/NewSensorStepThree.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/NewSensor/NewSensorStepThree.js]: 🙄 
92c24a6

[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/NewSensor/NewSensorStepThree.js]: I give up 
b648ad8

File: src/pages/SensorEditPage.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/SensorEditPage.js]: nice! 

648cfe6

I just did it in a commit here (Also the change to the hook is semi-unrelated but ensures that if you use a picker i.e. the time picker the input updates correctly ....)

**Review by mark-prins — APPROVED** [review]

**Pull request merged**